{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "66de9fb1",
   "metadata": {},
   "source": [
    "![image.png](./figures/nuclei_figure.png)\n",
    "# Code companion to:\n",
    "## High resolution mapping of the breast cancer tumor microenvironment using integrated single cell, spatial and in situ analysis of FFPE tissue\n",
    "Amanda Janesick, Robert Shelansky, Andrew Gottscho, Florian Wagner, Morgane Rouault, Ghezal Beliakoff, Michelli Faria de Oliveira, Andrew Kohlway, Jawad Abousoud, Carolyn Morrison, Tingsheng Yu Drennon, Syrus Mohabbat, Stephen Williams, 10x Development Teams, Sarah Taylor\n",
    "\n",
    "October 20, 2023\n",
    "\n",
    "Link to the paper: https://www.biorxiv.org/content/10.1101/2022.10.06.510405v1\n",
    "\n",
    "Link to the data: https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast\n",
    "\n",
    "Available for download from GitHub: https://github.com/10XGenomics/janesick_nature_comms_2023_companion\n",
    "![image.png](./figures/F1.large.jpg)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9314e604",
   "metadata": {},
   "source": [
    "### Quick start guide\n",
    "#### Install conda and setup the conda environment\n",
    "Conda installation instructions may be found [here](https://conda.io/projects/conda/en/latest/user-guide/install/index.html). If you choose to install miniconda, you will also need to install [jupyter](https://jupyter.org/) and its dependencies. \n",
    "\n",
    "The conda environment file, ```environment.yaml```, may be found in the [10xGenomics github repository](https://github.com/10XGenomics/janesick_nature_comms_2023_companion). To create the publication environment, run the following commands in the terminal. After creating the environment from within the open notebook, select the environment, ```xenium-publication-env```, from the ```kernels``` dropdown menu. A resource for managing conda environments may be found [here](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html). "
   ]
  },
  {
   "cell_type": "raw",
   "id": "5bc623fe",
   "metadata": {},
   "source": [
    "## make sure conda is installed (miniconda)\n",
    "conda --version \n",
    "\n",
    "## download yaml, ipynb, py files from github repo\n",
    "## https://github.com/10XGenomics/janesick_nature_comms_2023_companion\n",
    "\n",
    "## create conda conda_environment\n",
    "conda env create -f environment.yaml\n",
    "\n",
    "## activate conda environment\n",
    "conda activate xenium-publication-env\n",
    "\n",
    "## check you have the right conda environment\n",
    "conda info --envs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72cded69",
   "metadata": {},
   "source": [
    "### Publication data paths\n",
    "Data used in this document can be downloaded at https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast.  For this analysis we will use data from ```In Situ Sample 1, Replicate 1``` and ```Visium Spatial```. In addition, you will also need to download the  ```Cell_Barcode_Type_Matrices.xlsx``` found in ```In Situ Sample 1, Replicate 1``` under the heading ```Cell Type Annotations```. Data for the publication can also be obtained from [GEO:GSE243275](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE243275)."
   ]
  },
  {
   "cell_type": "raw",
   "id": "2a495e61",
   "metadata": {},
   "source": [
    "## In your $HOME directory\n",
    "## Download and unzip the Xenium output bundle\n",
    "mkdir xenium-publication\n",
    "cd xenium-publication\n",
    "\n",
    "mkdir xenium\n",
    "cd xenium\n",
    "curl -O https://cf.10xgenomics.com/samples/xenium/1.0.1/Xenium_FFPE_Human_Breast_Cancer_Rep1/Xenium_FFPE_Human_Breast_Cancer_Rep1_outs.zip\n",
    "unzip Xenium_FFPE_Human_Breast_Cancer_Rep1_outs.zip\n",
    "cd ..\n",
    "\n",
    "## Download and unzip Visium data\n",
    "mkdir visium\n",
    "cd visium\n",
    "curl -O https://cf.10xgenomics.com/samples/spatial-exp/2.0.0/CytAssist_FFPE_Human_Breast_Cancer/CytAssist_FFPE_Human_Breast_Cancer_filtered_feature_bc_matrix.h5\n",
    "curl -O https://cf.10xgenomics.com/samples/spatial-exp/2.0.0/CytAssist_FFPE_Human_Breast_Cancer/CytAssist_FFPE_Human_Breast_Cancer_spatial.tar.gz\n",
    "unzip CytAssist_FFPE_Human_Breast_Cancer_spatial.tar.gz\n",
    "cd ..\n",
    "\n",
    "cd .."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbe5cd81",
   "metadata": {},
   "source": [
    "#### Saving the cell_groups.csv file\n",
    "The ```Cell_Barcode_Type_Matrices.xlsx``` contains all the annotation for the paper. For this analysis we will only make use of the supervised labeling from Xenium Sample 1, Replicate 1. Use ```SaveAs``` to save the ```Xenium R1 Fig1-5 (supervised)``` tab as a CSV file. We will refer to it in the code as the ```cell_groups.csv```. Move the ```cell_groups.csv``` into the ```xenium-publication``` directory we just created. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f807a0b",
   "metadata": {},
   "source": [
    "#### Define data paths\n",
    "The following file paths assume that the xenium-publication was created as described above in the user's $HOME directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "576f112b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "path_to_visium_bundle = Path(\"~/xenium-publication/visium/\").expanduser()\n",
    "path_to_xenium_bundle = Path(\"~/xenium-publication/xenium/outs\").expanduser()\n",
    "path_to_xenium_celltype_labels = Path(\"~/xenium-publication/cell_groups.csv\").expanduser()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "459595f1",
   "metadata": {},
   "source": [
    "### Module Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b20c0d0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from collections import namedtuple, Counter\n",
    "\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "## Image manipulation and geometry\n",
    "from tifffile import imread\n",
    "from skimage.io import imread as skimread\n",
    "\n",
    "## Plotting imports\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.colors import LinearSegmentedColormap, Normalize, to_hex, Colormap\n",
    "from matplotlib.cm import ScalarMappable\n",
    "\n",
    "### Functions for registration of Xenium to Visium data and associated analyis and visualization\n",
    "## Visualization functions\n",
    "from companion_functions import (\n",
    "    hexlist_to_cmap,\n",
    "    polygons_coords_to_patch_collection,\n",
    "    plot_polygons,\n",
    "    hex_corner_offset,\n",
    "    polygon_corners,\n",
    "    celltypes,\n",
    "    celltypes,\n",
    "    hex_codes,\n",
    "    ctype_to_code_map,\n",
    "    ctype_hex_map,\n",
    "    ctype_cmap,\n",
    ")\n",
    "\n",
    "## Analysis functions\n",
    "from companion_functions import (\n",
    "    unique_encode,\n",
    "    get_xenium_to_morphology_transform_from_xenium_morphology_image,\n",
    "    get_xenium_capture_polygon_um,\n",
    "    transform_coordinates,\n",
    "    get_median_spot_to_spot_distance_from_centroids,\n",
    "    generate_space_filling_visium_polygons,\n",
    "    get_visium_capture_polygon,\n",
    "    __OUTSIDE_VISIUM_CAPTURE_AREA_BARCODE__,\n",
    "    bin_xenium_data_to_visium_spots,\n",
    "    generate_anndata_matrix_from_transcript_assignments,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa84b944",
   "metadata": {},
   "source": [
    "#### Outputs for Visium data\n",
    "For an overview of Visium output files, see [this page](https://www.10xgenomics.com/support/software/space-ranger/analysis/outputs/output-overview).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e48a6b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load Visium spot X gene matrix\n",
    "visium_adata = sc.read_10x_h5(path_to_visium_bundle / \"CytAssist_FFPE_Human_Breast_Cancer_filtered_feature_bc_matrix.h5\")\n",
    "\n",
    "## Load Visium spot position information\n",
    "visium_tissue_positions = (\n",
    "    pd.read_csv(path_to_visium_bundle / \"spatial\" / \"tissue_positions.csv\")\n",
    "    .set_index(\"barcode\")\n",
    "    .reindex(visium_adata.obs.index)\n",
    ")\n",
    "\n",
    "## Load Visium scale metadata\n",
    "## Contains scalefactors to transform between Visium full resolution and the high resolution images.\n",
    "with open(path_to_visium_bundle / \"spatial\" / \"scalefactors_json.json\") as file:\n",
    "    visium_scale_factors = json.load(file)\n",
    "\n",
    "## Load the Visium high resolution H&E for plotting\n",
    "visium_high_res_image = skimread(\n",
    "    path_to_visium_bundle / \"spatial\" / \"tissue_hires_image.png\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "623b70a6",
   "metadata": {},
   "source": [
    "#### IO for Xenium data\n",
    "Information about the Xenium output bundle may be found [here](https://www.10xgenomics.com/support/in-situ-gene-expression/documentation/steps/onboard-analysis/understanding-xenium-outputs).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e57652b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read in Xenium transcript data and decode the bytes column to string\n",
    "xenium_transcripts = pd.read_parquet(path_to_xenium_bundle / \"transcripts.parquet\")\n",
    "xenium_transcripts.feature_name = xenium_transcripts.feature_name.str.decode(\"utf-8\")\n",
    "\n",
    "## Read in the Xenium cell positions and cell metadata.\n",
    "xenium_cell_info = pd.read_csv(path_to_xenium_bundle / \"cells.csv.gz\").set_index(\n",
    "    \"cell_id\"\n",
    ")\n",
    "\n",
    "## Read in the Xenium morphology (DAPI) MIP image. Note: We are reading the image at the resolution of\n",
    "## the third pyramid level: full_resolution / 2 ** 3. We are using this image for plotting.\n",
    "## The Xenium imaging area is large, and the full resolution image is large enough to be slow to plot.\n",
    "## Working with a lower resolution image when possible is a convenience.\n",
    "pyramid_level = 3\n",
    "xenium_morphology_mip_at_pyramid_level = imread(\n",
    "    path_to_xenium_bundle / \"morphology_mip.ome.tif\", series=0, level=pyramid_level\n",
    ")\n",
    "\n",
    "## Read in the nucleus boundary polygon information. Xenium polygons are a plotting convenience and should not\n",
    "## be used for analysis. A nucleus mask is provided for analysis of Xenium segmentation.\n",
    "xenium_nucleous_boundaries_df = pd.read_parquet(\n",
    "    path_to_xenium_bundle / \"nucleus_boundaries.parquet\"\n",
    ")\n",
    "\n",
    "## Transformation matrix to transform from the Xenium coordinate system to the morphology image coordinate system\n",
    "## at pyramid level. Determined using the pixel size of the morphology image. \n",
    "xenium_to_morphology_at_pyramid_transformation = (\n",
    "    get_xenium_to_morphology_transform_from_xenium_morphology_image(\n",
    "        path_to_xenium_bundle / \"morphology_mip.ome.tif\", pyramid_level\n",
    "    )\n",
    ")\n",
    "\n",
    "## Cell type assignments as determined by label transfer from a matched FLEX dataset.\n",
    "## This is a supplemental file provided within the publication.\n",
    "## barcode X celltype\n",
    "xenium_celltype_df = pd.read_csv(path_to_xenium_celltype_labels).set_index(\"Barcode\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9b0f527",
   "metadata": {},
   "source": [
    "### The transformation matrix as determined by registration using Xenium Explorer\n",
    "The following registration matrix was generated using the Xenium Explorer registration tool.\n",
    "The full resolution Visium H&E image was loaded into Xenium Explorer's H&E registration workflow, then the transformation matrix was exported using the transformation export feature. It was directly copied into the cell below for convenience. Some further addendums to the transformation matrix are required to map Xenium coordinates into Visium coordinates. They are described inline. \n",
    "\n",
    "For information about the Xenium Explorer H&E registration workflow, see the [Xenium Explorer image alignment tutorial](\n",
    "https://www.10xgenomics.com/support/software/xenium-explorer/tutorials/xe-image-alignment)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ceb6167c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Transformation matrix exported from Xenium Explorer\n",
    "xenium_morphology_to_visium_full_res_transformation = np.array(\n",
    "    [\n",
    "        [1.87594468e-02, -4.07641677e-01, 1.63476055e04],\n",
    "        [3.91300071e-01, 1.26819551e-02, 4.12499099e03],\n",
    "        [-8.39854141e-08, -9.91112632e-07, 1.03706895e00],\n",
    "    ]\n",
    ")\n",
    "\n",
    "## Transformation matrix between Xenium full resolution (DAPI) morphology image and the Xenium coordinate system.\n",
    "xenium_to_morphology_transformation = (\n",
    "    get_xenium_to_morphology_transform_from_xenium_morphology_image(\n",
    "        path_to_xenium_bundle / \"morphology_mip.ome.tif\", 0\n",
    "    )\n",
    ")\n",
    "\n",
    "## Xenium coordinates must be transformed into image coordinates before being transformed into Visium image coordinates.\n",
    "## The following describes that combined transformation (a matrix multiplication of the two transformation matrices).\n",
    "xenium_to_visium_full_res_transformation = (\n",
    "    xenium_morphology_to_visium_full_res_transformation\n",
    "    @ xenium_to_morphology_transformation\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "818bc000",
   "metadata": {},
   "source": [
    "## Figures demonstrating Xenium / Visium registration and analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ecee6f5",
   "metadata": {},
   "source": [
    "#### Xenium coordinate transformations\n",
    "The following figure describes a transformation between the Xenium coordinate space and the Xenium morphology image. \n",
    "However, for plotting convenience, the target morphology image is not full resolution. \n",
    "\n",
    "\n",
    "The DAPI image is in the background, while gold cells centroids are overlaid as points. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "974df504",
   "metadata": {},
   "outputs": [],
   "source": [
    "## data setup\n",
    "## get Xenium coordinates -> transformed into Xenium morphology pixel coordinates\n",
    "coords = transform_coordinates(\n",
    "    xenium_cell_info[[\"x_centroid\", \"y_centroid\"]].values,\n",
    "    xenium_to_morphology_at_pyramid_transformation,\n",
    ")\n",
    "\n",
    "## plot style setup\n",
    "fig = plt.figure(figsize=(8, 8))\n",
    "ax = plt.gca()\n",
    "\n",
    "## plotting code\n",
    "ax.imshow(xenium_morphology_mip_at_pyramid_level)\n",
    "ax.scatter(*coords.T, s=0.1, linewidth=0, edgecolor=None, marker=\"o\", color=\"yellow\")\n",
    "\n",
    "## plot style adjustments\n",
    "ax.spines.right.set_visible(False)\n",
    "ax.spines.top.set_visible(False)\n",
    "ax.yaxis.set_ticks_position(\"left\")\n",
    "ax.xaxis.set_ticks_position(\"bottom\")\n",
    "ax.set_aspect(\"equal\", adjustable=\"box\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1c85240",
   "metadata": {},
   "source": [
    "#### Figure demonstrating the Xenium/Visium registration\n",
    "\n",
    "The Visium H&E high resolution image is displayed with Xenium cell coordinates transformed and overlaid in gold. For demonstration, the Xenium morphology image bounding polygon is displayed in blue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0c736e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## data setup\n",
    "## Xenium cell coordinates are transformed first into Visium full resolution image coordinates,\n",
    "## and then scaled to high resolution image coordinates.\n",
    "xenium_cell_centroids_visium_high_res = (\n",
    "    transform_coordinates(\n",
    "        xenium_cell_info[[\"x_centroid\", \"y_centroid\"]].values,\n",
    "        xenium_to_visium_full_res_transformation,\n",
    "    )\n",
    "    * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    ")\n",
    "\n",
    "## The Xenium capture polygon is obtained and similarly transformed.\n",
    "xenium_capture_polygon_um = get_xenium_capture_polygon_um(path_to_xenium_bundle)\n",
    "xenium_capture_polygon_visium_high_res = (\n",
    "    transform_coordinates(\n",
    "        xenium_capture_polygon_um, xenium_to_visium_full_res_transformation\n",
    "    )\n",
    "    * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    ")\n",
    "\n",
    "## plot style setup\n",
    "fig = plt.figure(figsize=(8, 8))\n",
    "ax = plt.gca()\n",
    "\n",
    "## plotting code\n",
    "ax.imshow(visium_high_res_image)\n",
    "ax.scatter(\n",
    "    *xenium_cell_centroids_visium_high_res.T,\n",
    "    s=0.1,\n",
    "    linewidth=0,\n",
    "    edgecolor=None,\n",
    "    marker=\"o\",\n",
    "    color=\"gold\"\n",
    ")\n",
    "plot_polygons(\n",
    "    [xenium_capture_polygon_visium_high_res],\n",
    "    ax=ax,\n",
    "    facecolor=(0, 0, 0, 0),\n",
    "    edgecolor=\"blue\",\n",
    "    linewidth=1,\n",
    ")\n",
    "\n",
    "## plot style adjustments\n",
    "ax.spines.right.set_visible(False)\n",
    "ax.spines.top.set_visible(False)\n",
    "ax.yaxis.set_ticks_position(\"left\")\n",
    "ax.xaxis.set_ticks_position(\"bottom\")\n",
    "ax.set_aspect(\"equal\", adjustable=\"box\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4079fad9",
   "metadata": {},
   "source": [
    "#### Figure demonstrating the Visium capture area and spot locations\n",
    "The Visium H&E high resolution image is displayed with the Visium capture area, displayed in cyan, and the Visium spots, represented as space-filling hexagons in black."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a92b75d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## data setup\n",
    "## Space filling polygons are generated and then transformed in the high resolution coordinate system.\n",
    "visium_hexegon_polygons = generate_space_filling_visium_polygons(\n",
    "    spot_coords=visium_tissue_positions[\n",
    "        [\"pxl_col_in_fullres\", \"pxl_row_in_fullres\"]\n",
    "    ].values,\n",
    ")\n",
    "\n",
    "visium_capture_polygon_full_res = get_visium_capture_polygon(\n",
    "    spot_centroids=visium_tissue_positions[\n",
    "        [\"pxl_col_in_fullres\", \"pxl_row_in_fullres\"]\n",
    "    ].values,\n",
    "    spot_diameter=visium_scale_factors[\"spot_diameter_fullres\"],\n",
    ")\n",
    "\n",
    "\n",
    "visium_hexagon_polygons_high_res = (\n",
    "    visium_hexegon_polygons * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    ")\n",
    "visium_capture_polygon_high_res = (\n",
    "    visium_capture_polygon_full_res * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    ")\n",
    "\n",
    "\n",
    "## plot style setup\n",
    "fig = plt.figure(figsize=(8, 8))\n",
    "ax = plt.gca()\n",
    "\n",
    "## plotting code\n",
    "ax.imshow(visium_high_res_image)\n",
    "plot_polygons(\n",
    "    visium_hexagon_polygons_high_res,\n",
    "    ax=ax,\n",
    "    facecolor=(0, 0, 0, 0),\n",
    "    edgecolor=\"black\",\n",
    "    linewidth=1,\n",
    ")\n",
    "plot_polygons(\n",
    "    [visium_capture_polygon_high_res],\n",
    "    ax=ax,\n",
    "    facecolor=(0, 0, 0, 0),\n",
    "    edgecolor=\"cyan\",\n",
    "    linewidth=1,\n",
    ")\n",
    "\n",
    "## plot style adjustments\n",
    "ax.spines.right.set_visible(False)\n",
    "ax.spines.top.set_visible(False)\n",
    "ax.yaxis.set_ticks_position(\"left\")\n",
    "ax.xaxis.set_ticks_position(\"bottom\")\n",
    "ax.set_aspect(\"equal\", adjustable=\"box\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68d81de1",
   "metadata": {},
   "source": [
    "#### Figure demonstrating the Visium data for the gene TACSTD2\n",
    "The Visium H&E high resolution image is displayed with the Visium capture area, displayed in red, and the Visium spots, represented as space-filling hexagons colored by the UMI count of TACSTD2 in each spot. The colormap is set such that the deepest red is set to the 95 percentile.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92fbf6a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Data setup\n",
    "gene = \"TACSTD2\"\n",
    "feature_barcode_df = visium_adata.to_df()\n",
    "\n",
    "## Space filling polygons are generated and then transformed in the high resolution coordinate system.\n",
    "hexegon_polygons_fullres = generate_space_filling_visium_polygons(\n",
    "    spot_coords=visium_tissue_positions.loc[feature_barcode_df.index][\n",
    "        [\"pxl_col_in_fullres\", \"pxl_row_in_fullres\"]\n",
    "    ].values,\n",
    ")\n",
    "hexegon_polygons_highres = (\n",
    "    hexegon_polygons_fullres * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    ")\n",
    "\n",
    "## UMI counts per barcode\n",
    "counts = feature_barcode_df.loc[:, gene].values\n",
    "\n",
    "## plot style setup\n",
    "fig = plt.figure(figsize=(8, 8))\n",
    "ax = plt.gca()\n",
    "norm = Normalize(vmin=0, vmax=np.quantile(counts, [0.95])[0])\n",
    "cmap = LinearSegmentedColormap.from_list(\n",
    "    \"Yellow/Orange/Red\", colors=[\"#ebebeb\", \"#FFD700\", \"#DC143C\", \"#4B0082\"]\n",
    ")\n",
    "\n",
    "## plotting code\n",
    "ax.imshow(visium_high_res_image)\n",
    "plot_polygons(hexegon_polygons_highres, ax=ax, array=counts, cmap=cmap, norm=norm)\n",
    "sm = ScalarMappable(cmap=cmap, norm=norm)\n",
    "fig.colorbar(sm, ax=ax)\n",
    "\n",
    "## plot style adjustments\n",
    "ax.spines.right.set_visible(False)\n",
    "ax.spines.top.set_visible(False)\n",
    "ax.yaxis.set_ticks_position(\"left\")\n",
    "ax.xaxis.set_ticks_position(\"bottom\")\n",
    "ax.set_aspect(\"equal\", adjustable=\"box\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87222c76",
   "metadata": {},
   "source": [
    "#### Figure demonstrating the different capture areas for the Xenium assay and the Visium assay run on two different slides. \n",
    "The Visium H&E high resolution image is displayed with the Visium capture area displayed in cyan and the Visium spots represented as space-filling hexagons colored by in black. Xenium cell coordinates are transformed and overlaid in gold. The Xenium morphology image bounding polygon is displayed in blue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db9065ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Data setup\n",
    "## Xenium cell cordinates are transformed into Visium high resolution image coordinates.\n",
    "xenium_cell_centroids_visium_high_res = (\n",
    "    transform_coordinates(\n",
    "        xenium_cell_info[[\"x_centroid\", \"y_centroid\"]].values,\n",
    "        xenium_to_visium_full_res_transformation,\n",
    "    )\n",
    "    * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    ")\n",
    "\n",
    "## Xenium capture area polygon is transformed into Visium high resolution image coordinates.\n",
    "xenium_capture_polygon_um = get_xenium_capture_polygon_um(path_to_xenium_bundle)\n",
    "xenium_capture_polygon_visium_high_res = (\n",
    "    transform_coordinates(\n",
    "        xenium_capture_polygon_um, xenium_to_visium_full_res_transformation\n",
    "    )\n",
    "    * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    ")\n",
    "\n",
    "## Space filling polygons are generated and then transformed  into high resolution image coordinates.\n",
    "visium_hexegon_polygons = generate_space_filling_visium_polygons(\n",
    "    spot_coords=visium_tissue_positions[\n",
    "        [\"pxl_col_in_fullres\", \"pxl_row_in_fullres\"]\n",
    "    ].values,\n",
    ")\n",
    "## Visium capture area polygons are generated and then transformed into high resolution image coordinates.\n",
    "visium_catpure_polygon_full_res = get_visium_capture_polygon(\n",
    "    spot_centroids=visium_tissue_positions[\n",
    "        [\"pxl_col_in_fullres\", \"pxl_row_in_fullres\"]\n",
    "    ].values,\n",
    "    spot_diameter=visium_scale_factors[\"spot_diameter_fullres\"],\n",
    ")\n",
    "\n",
    "visium_hexagon_polygons_high_res = (\n",
    "    visium_hexegon_polygons * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    ")\n",
    "visium_catpure_polygon_high_res = (\n",
    "    visium_catpure_polygon_full_res * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    ")\n",
    "\n",
    "## plot style setup\n",
    "fig = plt.figure(figsize=(8, 8))\n",
    "ax = plt.gca()\n",
    "\n",
    "## plotting code\n",
    "ax.imshow(visium_high_res_image)\n",
    "plot_polygons(\n",
    "    visium_hexagon_polygons_high_res,\n",
    "    ax=ax,\n",
    "    facecolor=(0, 0, 0, 0),\n",
    "    edgecolor=\"black\",\n",
    "    linewidth=1,\n",
    ")\n",
    "plot_polygons(\n",
    "    [visium_catpure_polygon_high_res],\n",
    "    ax=ax,\n",
    "    facecolor=(0, 0, 0, 0),\n",
    "    edgecolor=\"cyan\",\n",
    "    linewidth=1,\n",
    ")\n",
    "ax.scatter(\n",
    "    *xenium_cell_centroids_visium_high_res.T,\n",
    "    s=0.1,\n",
    "    linewidth=0,\n",
    "    edgecolor=None,\n",
    "    marker=\"o\",\n",
    "    color=\"gold\"\n",
    ")\n",
    "plot_polygons(\n",
    "    [xenium_capture_polygon_visium_high_res],\n",
    "    ax=ax,\n",
    "    facecolor=(0, 0, 0, 0),\n",
    "    edgecolor=\"blue\",\n",
    "    linewidth=1,\n",
    ")\n",
    "\n",
    "## plot style adjustments\n",
    "ax.spines.right.set_visible(False)\n",
    "ax.spines.top.set_visible(False)\n",
    "ax.yaxis.set_ticks_position(\"left\")\n",
    "ax.xaxis.set_ticks_position(\"bottom\")\n",
    "ax.set_aspect(\"equal\", adjustable=\"box\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f672b1d",
   "metadata": {},
   "source": [
    "### Visium spot interpolation of Xenium data using Xenium/Visium registration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89d3254a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter Xenium transcripts for only *high* quality transcripts. The Xenium cell-feature matrix is generated only\n",
    "## from high quality transcripts. However, the Xenium output bundle includes low quality transcripts. The following\n",
    "## interpolation is to be accomplished using the same transcripts found in the cell-feature matrix.\n",
    "## For more information see:\n",
    "## https://www.10xgenomics.com/resources/analysis-guides/continuing-your-journey-after-xenium-analyzer\n",
    "xenium_transcript_selection = xenium_transcripts.qv >= 20\n",
    "\n",
    "## Transcript to spot interpolation\n",
    "xenium_spotbinned_transcript_barcodes = bin_xenium_data_to_visium_spots(\n",
    "    visium_spot_centroids=visium_tissue_positions[\n",
    "        [\"pxl_col_in_fullres\", \"pxl_row_in_fullres\"]\n",
    "    ].values,\n",
    "    visium_spot_diameter=visium_scale_factors[\"spot_diameter_fullres\"],\n",
    "    visium_barcodes=visium_tissue_positions.index.values,\n",
    "    xenium_coords=xenium_transcripts[xenium_transcript_selection][\n",
    "        [\"x_location\", \"y_location\"]\n",
    "    ].values,\n",
    "    xenium_to_visium_transform=xenium_to_visium_full_res_transformation,\n",
    ")\n",
    "\n",
    "## The Xenium capture area does not perfectly overlap the Visium capture area. We want to\n",
    "## ignore Xenium transcripts which fall outside of the overlap.\n",
    "unassigned_transcript_mask = (\n",
    "    xenium_spotbinned_transcript_barcodes != __OUTSIDE_VISIUM_CAPTURE_AREA_BARCODE__\n",
    ")\n",
    "\n",
    "## Generating a cell-feature matrix from the interpolation\n",
    "xenium_spotbinned_transcript_adata = (\n",
    "    generate_anndata_matrix_from_transcript_assignments(\n",
    "        barcodes=xenium_spotbinned_transcript_barcodes[unassigned_transcript_mask],\n",
    "        feature_names=xenium_transcripts[xenium_transcript_selection][\n",
    "            unassigned_transcript_mask\n",
    "        ][\"feature_name\"].values,\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b9c20aa",
   "metadata": {},
   "source": [
    "#### Figure demonstrating Xenium spot interpolation of Visium spots for the gene TACSTD2\n",
    "The Visium H&E high resolution image is displayed with the interpolated Visium spots represented as space filling hexagons colored by the Xenium transcript counts for TACSTD2 assigned to each spot. The colormap is set such that the deepest red is set to the 95 percentile."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7bb25d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = \"TACSTD2\"\n",
    "feature_barcode_df = xenium_spotbinned_transcript_adata.to_df()\n",
    "\n",
    "## get Visium spot space-filling polygons and transform them into the high resolution image coordinates.\n",
    "hexegon_polygons_fullres = generate_space_filling_visium_polygons(\n",
    "    spot_coords=visium_tissue_positions.loc[feature_barcode_df.index][\n",
    "        [\"pxl_col_in_fullres\", \"pxl_row_in_fullres\"]\n",
    "    ].values,\n",
    ")\n",
    "hexegon_polygons_highres = (\n",
    "    hexegon_polygons_fullres * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    ")\n",
    "\n",
    "## collect gene counts for each spot\n",
    "counts = feature_barcode_df.loc[:, gene].values\n",
    "\n",
    "## plot style setup\n",
    "norm = Normalize(vmin=0, vmax=np.quantile(counts, [0.95])[0])\n",
    "cmap = LinearSegmentedColormap.from_list(\n",
    "    \"Yellow/Orange/Red\", colors=[\"#ebebeb\", \"#FFD700\", \"#DC143C\", \"#4B0082\"]\n",
    ")\n",
    "fig = plt.figure(figsize=(8, 8))\n",
    "ax = plt.gca()\n",
    "\n",
    "## plotting code\n",
    "ax.imshow(visium_high_res_image)\n",
    "plot_polygons(hexegon_polygons_highres, ax=ax, array=counts, cmap=cmap, norm=norm)\n",
    "sm = ScalarMappable(cmap=cmap, norm=norm)\n",
    "fig.colorbar(sm, ax=ax)\n",
    "\n",
    "## plot style adjustments\n",
    "ax.spines.right.set_visible(False)\n",
    "ax.spines.top.set_visible(False)\n",
    "ax.yaxis.set_ticks_position(\"left\")\n",
    "ax.xaxis.set_ticks_position(\"bottom\")\n",
    "ax.set_aspect(\"equal\", adjustable=\"box\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68f931fe",
   "metadata": {},
   "source": [
    "## Figure showing correlation between interpolated/actual Visium spots for the gene TACSTD2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7e7ffd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Data setup\n",
    "## Collating both interpolated Visium spots and actual Visium spots\n",
    "## The Xenium and Visium gene sets differ; here we are filtering the two data frames to\n",
    "## ensure they share the exact same genes. In addition, we can only interpolate Visium spot counts for\n",
    "## which we have Xenium data. We remove all spots which do not overlap between the two modalities.\n",
    "xenium_df = xenium_spotbinned_transcript_adata.to_df()\n",
    "visium_df = visium_adata.to_df()\n",
    "\n",
    "shared_barcodes = xenium_df.index.intersection(visium_df.index)\n",
    "shared_genes = xenium_df.columns.intersection(visium_df.columns)\n",
    "\n",
    "shared_xenium_df = xenium_df.loc[shared_barcodes, shared_genes]\n",
    "shared_visium_df = visium_df.loc[shared_barcodes, shared_genes]\n",
    "\n",
    "Gene = \"TACSTD2\"\n",
    "\n",
    "## Plot style setup\n",
    "fig = plt.figure(figsize=(8, 8))\n",
    "ax = plt.gca()\n",
    "\n",
    "## Plotting code\n",
    "ax.scatter(shared_visium_df[Gene], shared_xenium_df[Gene], s=5)\n",
    "ax.set_xscale(\"log\")\n",
    "ax.set_yscale(\"log\")\n",
    "\n",
    "## Plot style adjustments\n",
    "ax.set_xlabel(\"Xenium transcript count\")\n",
    "ax.set_ylabel(\"Visium transcript count\")\n",
    "ax.spines.right.set_visible(False)\n",
    "ax.spines.top.set_visible(False)\n",
    "ax.yaxis.set_ticks_position(\"left\")\n",
    "ax.xaxis.set_ticks_position(\"bottom\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cb78eb7",
   "metadata": {},
   "source": [
    "#### Figure showing Xenium data overlaid with label-transferred cell types\n",
    "Data wrangling the polygons is a bit slow, however it can be worth the wait, the resolution of the segmentation allows for beautiful whole tissue figures. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7f17ad2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Data setup\n",
    "## Collection Xenium nucleous boundaries from the parquet file. Polygons are organized in order and each row\n",
    "## corresponds to an individual vertex.\n",
    "## This line is the slow bit. Cache for faster plotting.\n",
    "xenium_nucleous_polygons = [\n",
    "    poly[[\"vertex_x\", \"vertex_y\"]].values\n",
    "    for i, poly in xenium_nucleous_boundaries_df.groupby(\"cell_id\")\n",
    "]\n",
    "\n",
    "## Plot style setup\n",
    "fig = plt.figure(figsize=(8, 8))\n",
    "ax = plt.gca()\n",
    "\n",
    "## Plotting code\n",
    "plot_polygons(\n",
    "    xenium_nucleous_polygons,\n",
    "    ax=ax,\n",
    "    array=list(map(ctype_to_code_map.get, xenium_celltype_df.Cluster.values)),\n",
    "    edgecolor=\"black\",\n",
    "    linewidth=0.01,\n",
    "    norm=Normalize(vmin=0, vmax=len(celltypes)),\n",
    "    cmap=ctype_cmap,\n",
    ")\n",
    "\n",
    "## Plot style adjustments\n",
    "ax.spines.right.set_visible(False)\n",
    "ax.spines.top.set_visible(False)\n",
    "ax.yaxis.set_ticks_position(\"left\")\n",
    "ax.xaxis.set_ticks_position(\"bottom\")\n",
    "ax.set_aspect(\"equal\", adjustable=\"box\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9339a19",
   "metadata": {},
   "source": [
    "## Figure showing Xenium cell type proportions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c069af44",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Data setup\n",
    "celltype_data = xenium_celltype_df.groupby(\"Cluster\").size()\n",
    "normed_celltype_data = (celltype_data / celltype_data.sum()).sort_values(\n",
    "    ascending=False\n",
    ")\n",
    "\n",
    "## Plot style setup\n",
    "fig = plt.figure(figsize=(3, 5))\n",
    "ax = plt.gca()\n",
    "\n",
    "## Plotting code\n",
    "x = [0]\n",
    "data = celltype_data.values[:, np.newaxis]\n",
    "bottom = np.zeros(data.shape[1])\n",
    "for ctype, y in zip(celltype_data.index.values, data):\n",
    "    ax.bar(x, y, bottom=bottom, color=ctype_hex_map[ctype])\n",
    "    bottom += y\n",
    "\n",
    "\n",
    "## Plot style adjustments\n",
    "ax.xaxis.set_ticks(x)\n",
    "ticks = ax.set_xticklabels([\"Replicate 1\"], rotation=45, ha=\"right\")\n",
    "ax.spines.right.set_visible(False)\n",
    "ax.spines.top.set_visible(False)\n",
    "ax.yaxis.set_ticks_position(\"left\")\n",
    "ax.xaxis.set_ticks_position(\"bottom\")\n",
    "\n",
    "text = ax.set_ylabel(\"Cells\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16724fe4",
   "metadata": {},
   "source": [
    "#### Figure demonstrating cell and cell type  interpolation of Visium data from Xenium data\n",
    "Similar to transcript counts, the Xenium cells themselves may be binned to Visium spots. The figure below overlays Xenium cells and their cell types on top of Visium space-filling polygons. The following dataset is composed of exactly this binning. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0d7a973",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Data setup\n",
    "## Collection of Xenium nucleus boundaries from the parquet file. Polygons are organized in order and each row\n",
    "## corresponds to an individual vertex.\n",
    "## This line is the slow bit. Cache for faster plotting.\n",
    "xenium_nucleous_polygons = [\n",
    "    poly[[\"vertex_x\", \"vertex_y\"]].values\n",
    "    for i, poly in xenium_nucleous_boundaries_df.groupby(\"cell_id\")\n",
    "]\n",
    "\n",
    "\n",
    "## Transform Xenium nucleus boundary polygons into the Visium high resolution coordinate space.\n",
    "xenium_nucleous_polygons_visium_high_res = [\n",
    "    transform_coordinates(poly, xenium_to_visium_full_res_transformation)\n",
    "    * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    "    for poly in xenium_nucleous_polygons\n",
    "]\n",
    "\n",
    "## Xenium capture area polygon is transformed into Visium high resolution image coordinates.\n",
    "xenium_capture_polygon_um = get_xenium_capture_polygon_um(path_to_xenium_bundle)\n",
    "xenium_capture_polygon_visium_high_res = (\n",
    "    transform_coordinates(\n",
    "        xenium_capture_polygon_um, xenium_to_visium_full_res_transformation\n",
    "    )\n",
    "    * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    ")\n",
    "\n",
    "## Space-filling polygons are generated and then transformed into high resolution image coordinates.\n",
    "visium_hexegon_polygons = generate_space_filling_visium_polygons(\n",
    "    spot_coords=visium_tissue_positions[\n",
    "        [\"pxl_col_in_fullres\", \"pxl_row_in_fullres\"]\n",
    "    ].values,\n",
    ")\n",
    "\n",
    "\n",
    "## Visium capture area polygons are generated and then transformed into high resolution image coordinates.\n",
    "visium_catpure_polygon_full_res = get_visium_capture_polygon(\n",
    "    spot_centroids=visium_tissue_positions[\n",
    "        [\"pxl_col_in_fullres\", \"pxl_row_in_fullres\"]\n",
    "    ].values,\n",
    "    spot_diameter=visium_scale_factors[\"spot_diameter_fullres\"],\n",
    ")\n",
    "\n",
    "\n",
    "visium_hexagon_polygons_high_res = (\n",
    "    visium_hexegon_polygons * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    ")\n",
    "visium_catpure_polygon_high_res = (\n",
    "    visium_catpure_polygon_full_res * visium_scale_factors[\"tissue_hires_scalef\"]\n",
    ")\n",
    "\n",
    "\n",
    "## plot style setup\n",
    "fig = plt.figure(figsize=(8, 8))\n",
    "ax = plt.gca()\n",
    "\n",
    "\n",
    "## Plotting code\n",
    "ax.imshow(visium_high_res_image)\n",
    "plot_polygons(\n",
    "    visium_hexagon_polygons_high_res,\n",
    "    ax=ax,\n",
    "    facecolor=(0, 0, 0, 0),\n",
    "    edgecolor=\"black\",\n",
    "    linewidth=1,\n",
    ")\n",
    "plot_polygons(\n",
    "    [visium_catpure_polygon_high_res],\n",
    "    ax=ax,\n",
    "    facecolor=(0, 0, 0, 0),\n",
    "    edgecolor=\"cyan\",\n",
    "    linewidth=1,\n",
    ")\n",
    "\n",
    "plot_polygons(\n",
    "    xenium_nucleous_polygons_visium_high_res,\n",
    "    ax=ax,\n",
    "    array=list(map(ctype_to_code_map.get, xenium_celltype_df.Cluster.values)),\n",
    "    edgecolor=\"black\",\n",
    "    linewidth=0.01,\n",
    "    norm=Normalize(vmin=0, vmax=len(celltypes)),\n",
    "    cmap=ctype_cmap,\n",
    ")\n",
    "\n",
    "plot_polygons(\n",
    "    [xenium_capture_polygon_visium_high_res],\n",
    "    ax=ax,\n",
    "    facecolor=(0, 0, 0, 0),\n",
    "    edgecolor=\"blue\",\n",
    "    linewidth=1,\n",
    ")\n",
    "\n",
    "## Plot style adjustments\n",
    "ax.spines.right.set_visible(False)\n",
    "ax.spines.top.set_visible(False)\n",
    "ax.yaxis.set_ticks_position(\"left\")\n",
    "ax.xaxis.set_ticks_position(\"bottom\")\n",
    "ax.set_aspect(\"equal\", adjustable=\"box\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45f67416",
   "metadata": {},
   "source": [
    "### Visium spot interpolation of Xenium cell data using Xenium/Visium registration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "315c80a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get cell -> Visium barcode assignments.\n",
    "xenium_spotbinned_cell_barcodes = bin_xenium_data_to_visium_spots(\n",
    "    visium_spot_centroids=visium_tissue_positions[\n",
    "        [\"pxl_col_in_fullres\", \"pxl_row_in_fullres\"]\n",
    "    ].values,\n",
    "    visium_spot_diameter=visium_scale_factors[\"spot_diameter_fullres\"],\n",
    "    visium_barcodes=visium_tissue_positions.index.values,\n",
    "    xenium_coords=xenium_cell_info[[\"x_centroid\", \"y_centroid\"]].values,\n",
    "    xenium_to_visium_transform=xenium_to_visium_full_res_transformation,\n",
    ")\n",
    "\n",
    "## Filter cells outside of Visium capture area.\n",
    "unassigned_cell_mask = (\n",
    "    xenium_spotbinned_cell_barcodes != __OUTSIDE_VISIUM_CAPTURE_AREA_BARCODE__\n",
    ")\n",
    "\n",
    "## Generate a spot X cell type dataframe.\n",
    "xenium_spotbinned_cell_adata = generate_anndata_matrix_from_transcript_assignments(\n",
    "    barcodes=xenium_spotbinned_cell_barcodes[unassigned_cell_mask],\n",
    "    feature_names=xenium_celltype_df[unassigned_cell_mask][\"Cluster\"].values,\n",
    ")\n",
    "xenium_spotbinned_celltypes_df = xenium_spotbinned_cell_adata.to_df()\n",
    "\n",
    "## Normalize each spot to the sum of the cell count. \n",
    "xenium_spotbinned_celltypes_normed_df = xenium_spotbinned_celltypes_df.div(\n",
    "    xenium_spotbinned_celltypes_df.sum(1), axis=0\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c675352",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "xenium_spotbinned_celltypes_normed_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xenium-publication-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
